lib/protobuf/json/decode.ex

```diff
@@ -0,0 +1,203 @@
+    for {oneof, index} <- oneofs,
+        {_field_num, %{oneof: ^index} = prop} <- field_props,
+        value = field_value(prop, data),
+        reduce: %{} do
+      %{^oneof => _duplicated} -> throw({:duplicated_oneof, oneof})
+      acc -> Map.put(acc, oneof, {prop.name_atom, decode_value(prop, value)})
+    end
+  end
+
+  defp field_value(%{json_name: json_key, name: name_key}, data) do
+    case data do
+      %{^json_key => value} -> value
+      %{^name_key => value} -> value
+      _ -> nil
+    end
+  end
+
+  defp decode_value(%{map?: true} = prop, map), do: decode_map(prop, map)
+  defp decode_value(%{repeated?: true} = prop, list), do: decode_repeated(prop, list)
+  defp decode_value(%{repeated?: false} = prop, value), do: decode_singular(prop, value)
+
+  defp decode_map(%{type: module, name_atom: field}, map) when is_map(map) do
+    %{field_props: field_props, field_tags: field_tags} = Utils.message_props(module)
+    key_type = field_props[field_tags[:key]].type
+    val_prop = field_props[field_tags[:value]]
+
+    for {key, val} <- map, into: %{} do
+      {decode_key(key_type, key, field), decode_singular(val_prop, val)}
+    end
+  end
+
+  defp decode_map(prop, bad_map), do: throw({:bad_map, prop.name_atom, bad_map})
+
+  defp decode_key(type, key, field) when is_binary(key) do
+    case parse_key(type, key) do
+      {:ok, decoded} -> decoded
+      :error -> throw({:bad_map_key, field, type, key})
+    end
+  end
+
+  defp decode_key(type, key, field), do: throw({:bad_map_key, field, type, key})
+
+  # Map keys can be of any scalar type except float, double and bytes. they
+  # must always be wrapped in strings. Other types should not compile.
+  defp parse_key(:string, key), do: {:ok, key}
+  defp parse_key(:bool, "true"), do: {:ok, true}
+  defp parse_key(:bool, "false"), do: {:ok, false}
+  defp parse_key(type, key) when type in @int_types, do: parse_int(key)
+  defp parse_key(_type, _key), do: :error
+
+  defp decode_repeated(prop, value) when is_list(value) do
+    for val <- value, do: decode_singular(prop, val)
+  end
+
+  defp decode_repeated(prop, value) do
+    throw({:bad_repeated, prop.name_atom, value})
+  end
+
+  defp decode_singular(%{type: :string} = prop, value) do
+    if is_binary(value),
+      do: value,
+      else: throw({:bad_string, prop.name_atom, value})
+  end
+
+  defp decode_singular(%{type: :bool} = prop, value) do
+    if is_boolean(value),
+      do: value,
+      else: throw({:bad_bool, prop.name_atom, value})
+  end
+
+  defp decode_singular(%{type: type} = prop, value) when type in @int_types do
+    with {:ok, integer} <- decode_integer(value),
+         true <- integer in @int_ranges[type] do
+      integer
+    else
+      _ -> throw({:bad_int, prop.name_atom, value})
+    end
+  end
+
+  defp decode_singular(%{type: type} = prop, value) when type in @float_types do
+    # TODO: validate int min-max range
```

Comment:
`double` fields will be automatically range-limited, because they are unrepresentable in Erlang already:

```elixir
iex> Jason.decode ~S|{"topSpeed":1.7976931348623159e+308}|
{:error,
 %Jason.DecodeError{
   data: "{\"topSpeed\":1.7976931348623159e+308}",
   position: 12,
   token: "1.7976931348623159e+308"
 }}
```

`float` fields however won't be automatically range-limited.